File: Cpp/N-Queens-II.java
class Solution {
    /**
     * Calculate the total number of distinct N-Queen solutions.
     * @param n: The number of queens.
     * @return: The total number of distinct solutions.
     */
    private int num;
    public void helper( int[][] vis, int pos, int n) {
        if (pos == n) {
            num++;
            return;
        }
        for (int i = 0; i < n; i++) {
            if (vis[0][i] == 0 && vis[1][pos + i] == 0 && vis[2][n - i + pos] == 0) {
                vis[0][i] = vis[1][pos + i] = vis[2][n - i + pos] = 1;
                helper(vis, pos + 1, n);
                vis[0][i] = vis[1][pos + i] = vis[2][n - i + pos] = 0;
            }
        }
    }
    public int totalNQueens(int n) {
        //write your code here
        num = 0;
        if (n == 0) return 0;
        int[][] vis = new int[3][30];
        helper(vis, 0, n);
        return num;
    }
};

File: Cpp/N-Queens-II.cpp
#include <vector>
#include <string>

class Solution {
public:
    /**
     * Get all distinct N-Queen solutions
     * @param n: The number of queens
     * @return: All distinct solutions
     * For example, A string '...Q' shows a queen on forth position
     */
    void dfs(int &res, vector<vector<int> > &vis, int pos, int n) {
        if (pos == n) {
            res ++;
            return;
        }
        for (int i = 0; i < n; i ++) {
            if (vis[0][i] == 0 && vis[1][i+pos] == 0 && vis[2][n-i+pos] == 0) {
                vis[0][i] = vis[1][i+pos] = vis[2][n-i+pos] = 1;
                dfs(res, vis, pos + 1, n);
                vis[0][i] = vis[1][i+pos] = vis[2][n-i+pos] = 0;
            }
        }
    }
    int totalNQueens(int n) {
        // write your code here
        vector<vector<int> > vis(3, vector<int>(30));
        int res = 0;
        dfs(res, vis, 0, n);
        return res;
    }
};

